### ISSUE TITLE
Request: Save Multiple Accounts in the config.json file and allow users to switch between accounts

### ISSUE THREAD

**Issue by wiscow** [ClassicUO/ClassicUO#1193]
Currently there is an option in the Login screen to "Save Account".  This will store the last account that was logged in.  Looking in the config.json it appears that it only saves the last account that was used to login.  Would it be feasible to abstract this information into an array in the json file and then auto-fill the password when the username matches?

I would be willing to implement this if it is a PR you would be willing to entertain.

**Comment by andreakarasho** [comment]
Forget to deal with settings.json, it's deprecated.
Instead @SaschaKP told me a way to store passwords in a different file. We can make a very cool thing if we mix these ideas.

### The idea
Basically what you want is a set of Usernames into a combobox linked with their passwords. Accounts will depends by the server address like:

```toml
[server-1]
usernames=[ "xxx", "yyy", "zzz", "abc" ]
passwords=[ "xxx_pas", "yyy_pas", "zzz_pas", "abc_pas" ]

[server-2]
usernames=[ "ppp", "ooo", "iii", "uuu" ]
passwords=[ "ppp_pas", "ooo_pas", "iii_pas", "uuu_pas" ]
```

So if your config points to `server-1`, the login screen will display ``` "xxx", "yyy", "zzz", "abc" ```  only.

### Howto
This is the hardest part.
Probably the best option is to store accouns and password when:
- [x] Save Account is checked
- [x] username & password fields are not blank

Then add a button to `Remove` the selected account to delete the entry from the accounts file.

**Comment by wiscow** [comment]
Very interesting.  I just cloned the repository last night.  So I'll have to dig into it a bit.  
would the combobox be in place of the Username input entry at the login?  

I imagine it would have to be a combination of text field and combo-box.  If this is the case, is there a component that exists today that allows for this behavior that we could model the input off of?

**Comment by andreakarasho** [comment]
No it does not exist at the moment.
Before adding components or doing PR it's better to plan the correct behavior

**Comment by wiscow** [comment]
yeah that sounds good to me.  I will need to spend a little bit of time reviewing how everything works and i can put together some ideas. In the meantime, if you have any additional ideas or can point me in a specific direction.  I am at your discretion.

**Comment by wiscow** [comment]
OK i got the solution up and running and was able to launch and take a look at what is going on.  Here are my thoughts:

The easiest way i see doing this is just based on the current login screen as it exists today.

essentially the process would go like this:

1.  At login prompt the user types their account name or it is pulled from an account file as the Last Account. Similar to the settings.json setup that is currently used.

2. when the text is entered the process would look at the account list to see if an existing account matches the one provided.  

3. If the text of the account textbox matches, auto-populate the password field accordingly.

If no account currently exists, and the user chooses the option to "Save Account", the password would be saved for that account.  

If an account currently exists, the user modifies the password, and has the "Save Account" flag checked, the password would be updated in the accounts file.

If the user does not have the "Save Accounts" flag checked, no information is saved into the accounts file.

The one downfall i see to this approach would be if a user incorrectly enters an account and tries to login the information would be saved in error.  However, i think that is easily circumvented by only saving once the login is successful. 

The user would also not have any way to delete the saved data which may be some kind of security risk I suppose.

The option you laid out would work as well.  However, like you mentioned there does not appear to be a component that would fit this model directly and would need to be created.  I would see this component as a combination of textbox and combobox that allows the user to type input, but when the drop down arrow is clicked the options are shown and when selected update the input text accordingly.  

I think in both of these scenarios the `Remove` functionality is going to be tricky without overwhelming the UI.

I would be OK with going in either direction.

**Comment by wiscow** [comment]
@andreakarasho 

## Propsal 

### Create a New Combo/Textbox component `stbTextBoxCombo`

This new component would inherit from `stbTextBox`.  If you are more akin to composition over inheritance this functionality could be easily baked into the `stbTextBox` component as well.  In order to make this work with the drawer of the combobox, we will need to pull the `ComboboxContextMenu` out of `Comboox` and make it `internal` as opposed to `private`. I currently created a copy of the class.  However, it will need a little bit of refactoring since it takes in a `ComboBox` as a reference.  I would suggest an interface here. I took a stab at it and i have a working component that uses pretty much all the functionality of `stbTextBox` and `CombooxContextMenu` with mostly coordinate changes (the comboox has funky placement). 

Here are some of the things i have implemented in this control:

- I Added the "dropdown" arrow to the inherited textbox and made it responsive by swiping the graphics from the scrollbar. when you click on the arrow it depresses and when you let up it goes back to the original image.

- I made it so that the arrow has a hit box.  Only clicking the arrow will cause the dropdown context menu to open.

- When an item is chosen from the context menu, the text is input into the Textbox, the caret moves to the end, and the textbox gains focus.

- If there are no items passed into the component, the arrow does not render and it acts as a normal `stbTextBox`.

### Add Account Control to the `ProfileManager.cs` class

In order to abstract the account management.  I would suggest putting this in the `ProfileManager.cs` class.  I know there is other discussion in this issue https://github.com/andreakarasho/ClassicUO/issues/1162 about security.  The abstraction would take away the need for a permanent storage location and if needed could be stored in a os user folder.  My personal opinion would be to add a `.json` file in the `Data/Profiles` folder.  This seems to make sense hierarchically. The file would contain the servers, accounts and encrypted passwords.

The data would be loaded when the LoginGump renders and the passwords and characters would be saved upon successful login in the `ServerListReceived` event in `LoginScene.cs` this would ensure only successful logins would be saved.  The save would only take place if the `Save Account` flag is checked.

### Unknowns

I'm not sure where a `delete` button would go.  I would imagine that this functionality would also use the ProfileManager.cs and just remove the username and password that match the username that is in the textbox at the time the button is pressed.  

I am not up to speed on the rendering side of all this.  I am sure there are gotchas with component placement and rendering that i am fully not aware of.

### Disclaimer

I am completely new to this.  I have a lot of programming experience but this is really my first dive into open source. 

Should i fork the repository and push my changes there or is it OK to push directly to this repository?

**Comment by andreakarasho** [comment]
Sounds good for me.
Clone the dev branch and Make a PR to it with these changes

**Comment by SaschaKP** [comment]
this is what I was thinking, in the reading process, to populate the new combobox, use the OrderedDictionary, since combobox is using an array, so ordered, the dictionary (part of ClassicUO.Utility.Collections) is especially useful to maintain list of usernames + password, I have some more things I have thinked about this, but it will require, from my side, some more thinking, to make it more logical and error-free, just to explain after the reading, what we will need is:
a server class or struct object, with IEquatable<T> and custom Hashcode to allow for server search into the dictionary, so we will get something like this:
lookup the server hostname if we are on DNS, save the resulting ip into conversion, when searching for server we look for both dns and ip, if we have only IP we find the matching ip we have, port is ignored (but we can check that too, maybe?), then we get the ordereddictionary of username + password strings, the ordered dictionary inside ClassicUO.Utility.Collections is to be used, since the one present into system.collections.specialized isn't usable to do this.
if the username is present into the dictionary we have a prepopulated password, otherwise password will be blank and we can opt in to save that account, if we select an existing account, the save flag is already checked, if unchecked the account will be removed from the dictionary and not save, AFTER successful login
more is in my mind, but as the thing will more forward, I'll add it into it, but this should be a correct way to handle it (always into user LOCAL directory, so only accessible by the actual OS user)

**Comment by andreakarasho** [comment]
I propose to store the IP only as `key`. DNS or not the address is unique as the sample i posted above:

(it's TOML, but can be resumed in JSON too)

```toml
[server-1]
usernames=[ "xxx", "yyy", "zzz", "abc" ]
passwords=[ "xxx_pas", "yyy_pas", "zzz_pas", "abc_pas" ]

[server-2]
usernames=[ "ppp", "ooo", "iii", "uuu" ]
passwords=[ "ppp_pas", "ooo_pas", "iii_pas", "uuu_pas" ]
```

**Comment by wiscow** [comment]
@andreakarasho @SaschaKP 

*VERY* Rough draft at: https://github.com/wiscow/ClassicUO/tree/multi-account-login

**Comment by wiscow** [comment]
@SaschaKP You read my mind on the delete portion. I was thinking the same thing.  If the un-check the save account flag, it would delete from the file.  That way we don't need to add another button to the login screen.

**Comment by wiscow** [comment]
if the profile directory moves according to: https://github.com/andreakarasho/ClassicUO/pull/1180, the code here would have to get slightly modified to support the new profile path.  However,  I think it would be a one-liner.

**Comment by wiscow** [comment]
Draft PR created: https://github.com/andreakarasho/ClassicUO/pull/1200
